Remove special gcc check

adah1972 opened the pull request:
The current code is tested to work with major GCC versions (from GCC 7 on).

This will close issue #222.

adah1972 committed:
Remove macro _MDSPAN_USE_ALIAS_TEMPLATE_ARGUMENT_DEDUCTION

As it is no longer used.

adah1972 committed:
Remove the special GCC version check

The current code works with GCC 7 and later.

mhoemmen commented on include/experimental/__p0009_bits/config.hpp:
```diff
@@ -198,23 +198,12 @@ static_assert(_MDSPAN_CPLUSPLUS >= MDSPAN_CXX_STD_14, "mdspan requires C++14 or
-#  if (defined(_MDSPAN_COMPILER_CLANG) || !defined(__GNUC__) || __GNUC__ >= 11) \
-      && ((defined(__cpp_deduction_guides) && __cpp_deduction_guides >= 201703) \
-         || (!defined(__cpp_deduction_guides) && MDSPAN_HAS_CXX_17))
+#  if (defined(__cpp_deduction_guides) && __cpp_deduction_guides >= 201703) \
```
@adah1972 Thank you for the contribution!  Have you tested this with GCC 8?  I've found that GCC 8 sometimes doesn't build C++17 code that works fine with GCC 7.5 or GCC 9.  The automated PR tests don't include a GCC 8 test.

adah1972 replied:
> @adah1972 Thank you for the contribution! Have you tested this with GCC 8? I've found that GCC 8 sometimes doesn't build C++17 code that works fine with GCC 7.5 or GCC 9. The automated PR tests don't include a GCC 8 test.

I am happy to report that all tests compile and pass with the current patch under GCC 8.5.0.

mhoemmen replied:
Thanks for testing this!

mhoemmen reviewed (approved):
Thanks for the contribution and tests!

crtrott commented:
hold off this fails with NVCC

crtrott commented:
(we effectively implicitly disabled CTAD for most CUDA versions because most didn't support GCC11 anyway).

crtrott committed:
Disable CTAD tests for layout mapping for NVCC

doesn't work up through NVCC 12

crtrott committed:
Disable CTAD again for NVCC older than 11.7

Crashes with internal compiler error. Tested 11.2, 11.4, 11.6 with gcc
9/10

/ascldap/users/crtrott/Kokkos/mdspan/include/experimental/__p0009_bits/mdspan.hpp:
In member function 'virtual void
TestMdspanCTAD_extents_pack_Test::TestBody()':
/ascldap/users/crtrott/Kokkos/mdspan/include/experimental/__p0009_bits/mdspan.hpp:78:25:
internal compiler error: in tsubst, at cp/pt.c:14560
   78 |   MDSPAN_INLINE_FUNCTION static constexpr size_t rank() noexcept
{ return extents_type::rank(); }

mhoemmen commented on include/experimental/__p0009_bits/config.hpp:
```diff
@@ -198,23 +198,13 @@ static_assert(_MDSPAN_CPLUSPLUS >= MDSPAN_CXX_STD_14, "mdspan requires C++14 or
-#  if (defined(_MDSPAN_COMPILER_CLANG) || !defined(__GNUC__) || __GNUC__ >= 11) \
-      && ((defined(__cpp_deduction_guides) && __cpp_deduction_guides >= 201703) \
-         || (!defined(__cpp_deduction_guides) && MDSPAN_HAS_CXX_17))
+#  if (!defined(__NVCC__) || (__CUDACC_VER_MAJOR__ >= 11 && __CUDACC_VER_MINOR__ >= 7)) && \
```
Looks good! : - )

mhoemmen replied:
@youyu3 FYI

mhoemmen commented on tests/test_layout_ctors.cpp:
```diff
@@ -203,7 +203,8 @@ TEST(TestLayoutLeftListInitialization, test_layout_left_extent_initialization) {
 }
 
-#if defined(_MDSPAN_USE_CLASS_TEMPLATE_ARGUMENT_DEDUCTION)
+// FIXME: CUDA NVCC including 12.0 does not like CTAD on nested classes
```
It's not really an mdspan FIXME; it's an NVCC FIXME ; - )  Would you consider rewording?

```suggestion
// CUDA NVCC, including 12.0, implements CTAD but
// is not always able to perform CTAD on nested classes.
```

Another option would be to add a macro to config.hpp specifically for this case.  Then, you could comment the macro there once, so you wouldn't have to repeat the comment for each disabled test.

crtrott replied:
The FIXME is just a reminder that we need to revisit this with later versions of CUDA.
